src/lib/optimizer/strategy/null_scan_removal_rule.cpp

```diff
@@ -0,0 +1,84 @@
+#include "null_scan_removal_rule.hpp"
+
+#include <algorithm>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "constant_mappings.hpp"
+#include "cost_estimation/abstract_cost_estimator.hpp"
+#include "expression/is_null_expression.hpp"
+#include "expression/lqp_column_expression.hpp"
+#include "hyrise.hpp"
+#include "logical_query_plan/abstract_lqp_node.hpp"
+#include "logical_query_plan/logical_plan_root_node.hpp"
+#include "logical_query_plan/lqp_utils.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/stored_table_node.hpp"
+#include "optimizer/join_ordering/join_graph.hpp"
+#include "statistics/cardinality_estimation_cache.hpp"
+#include "statistics/cardinality_estimator.hpp"
+#include "statistics/table_statistics.hpp"
+#include "utils/assert.hpp"
+
+namespace opossum {
+
+void NullScanRemovalRule::apply_to_plan(const std::shared_ptr<LogicalPlanRootNode>& root) const {
+  Assert(root->type == LQPNodeType::Root, "NullScanRemovalRule needs root to hold onto");
+
+  std::vector<std::shared_ptr<AbstractLQPNode>> nodes_to_remove;
+
+  // To determine if the rule applies to a node it must meet all of the following conditions:
+  // 1. The node must be of type Predicate
+  // 2. The predicate must be a `null expression`
+  // 3. The predicate condition must be `is not null`
+  // 4. The predicate operand needs to be an `LQP Column expression`
+  // 5. The original node of the `LQP Column expression` needs to be a not defective `storage table node`
+  // 6. The column (referenced by the `LQP Column expression`) is not nullable
+  // All nodes where the conditions apply are removed from the LQP Tree.
+  auto visiter = [&](const auto& node) {
+    // Checks condition 1
+    if (node->type != LQPNodeType::Predicate) return LQPVisitation::VisitInputs;
+
+    const auto predicate_node = std::dynamic_pointer_cast<PredicateNode>(node);
+    const auto predicate = predicate_node->predicate();
+    // Checks for condition 2
+    if (const auto is_null_expression = std::dynamic_pointer_cast<IsNullExpression>(predicate)) {
+      // Checks for condition 3
+      if (is_null_expression->predicate_condition == PredicateCondition::IsNull) return LQPVisitation::VisitInputs;
+
+      const auto& column = std::dynamic_pointer_cast<LQPColumnExpression>(is_null_expression->operand());
+      // Checks for condition 4
```

Comment:
Suggested grouping with newlines:

```
// Check for condition 3
(code)
(code)

// Check for condition 4
```

Reply:
Do you mean to change the code the following way?

```cpp
if (is_null_expression->predicate_condition == PredicateCondition::IsNull){
    return LQPVisitation::VisitInputs;
}
```

Reply:
No, I mean that I would put the comment first, then all steps needed for that condition, then a newline, and finally the next step/condition. Really minor, but it might help with understanding the structure.